core/trino-main/src/main/java/io/trino/metadata/StaticCatalogStore.java

```diff
@@ -27,47 +28,57 @@
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
+import java.util.concurrent.Callable;
+import java.util.concurrent.Executor;
 import java.util.concurrent.atomic.AtomicBoolean;
 
 import static com.google.common.base.MoreObjects.firstNonNull;
 import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.collect.ImmutableList.toImmutableList;
 import static io.airlift.configuration.ConfigurationLoader.loadPropertiesFrom;
+import static io.trino.util.Executors.executeUntilFailure;
+import static java.util.Objects.requireNonNull;
 
 public class StaticCatalogStore
 {
     private static final Logger log = Logger.get(StaticCatalogStore.class);
     private final ConnectorManager connectorManager;
     private final File catalogConfigurationDir;
     private final Set<String> disabledCatalogs;
+    private final Executor executor;
     private final AtomicBoolean catalogsLoading = new AtomicBoolean();
 
     @Inject
-    public StaticCatalogStore(ConnectorManager connectorManager, StaticCatalogStoreConfig config)
+    public StaticCatalogStore(ConnectorManager connectorManager, StaticCatalogStoreConfig config, @ForStartup Executor executor)
     {
         this(connectorManager,
                 config.getCatalogConfigurationDir(),
-                firstNonNull(config.getDisabledCatalogs(), ImmutableList.of()));
+                firstNonNull(config.getDisabledCatalogs(), ImmutableList.of()),
+                executor);
     }
 
-    public StaticCatalogStore(ConnectorManager connectorManager, File catalogConfigurationDir, List<String> disabledCatalogs)
+    public StaticCatalogStore(ConnectorManager connectorManager, File catalogConfigurationDir, List<String> disabledCatalogs, Executor executor)
     {
         this.connectorManager = connectorManager;
         this.catalogConfigurationDir = catalogConfigurationDir;
         this.disabledCatalogs = ImmutableSet.copyOf(disabledCatalogs);
+        this.executor = requireNonNull(executor, "executor is null");
     }
 
     public void loadCatalogs()
-            throws Exception
     {
         if (!catalogsLoading.compareAndSet(false, true)) {
             return;
         }
 
-        for (File file : listFiles(catalogConfigurationDir)) {
-            if (file.isFile() && file.getName().endsWith(".properties")) {
-                loadCatalog(file);
-            }
-        }
+        executeUntilFailure(
+                executor,
+                listFiles(catalogConfigurationDir).stream()
+                        .map(file -> (Callable<?>) () -> {
```

Comment:
Connectors and possibly other plugins tend to use and manipulate static resources like system properties. I think we might need to have two separate flags to load class concurrently and catalogs concurrently in separate.